File: packages/python/elizaos/advanced_capabilities/experience/utils/experience_analyzer.py
from __future__ import annotations

import random
from dataclasses import dataclass
from datetime import UTC, datetime

from elizaos.advanced_capabilities.experience.types import (
    Experience,
    ExperienceType,
    OutcomeType,
)


@dataclass
class ExperienceAnalysisResult:
    is_significant: bool
    learning: str | None = None
    confidence: float = 0.5
    related_experiences: list[str] | None = None
    actionable_insights: list[str] | None = None


@dataclass
class FailurePattern:
    learning: str
    related_ids: list[str]
    insights: list[str]


@dataclass
class DetectedPattern:
    description: str
    frequency: int
    experiences: list[str]
    significance: str  # "low" | "medium" | "high"


async def analyze_experience(
    partial_experience: dict[str, object],
    recent_experiences: list[Experience],
) -> ExperienceAnalysisResult:
    """Check if this experience represents something new or significant."""

    similar = _find_similar_experiences(partial_experience, recent_experiences)

    # If we've seen this exact pattern many times, it's less significant
    if len(similar) > 5:
        return ExperienceAnalysisResult(is_significant=False, confidence=0.3)

    # Check for contradictions with previous experiences
    contradictions = _find_contradictions(partial_experience, recent_experiences)
    if contradictions:
        first = contradictions[0]
        current_result = partial_experience.get("result", "unknown result")
        previous_result = first.result or "unknown result"

        return ExperienceAnalysisResult(
            is_significant=True,
            learning=f"New outcome contradicts previous experience: {current_result} vs {previous_result}",
            confidence=0.8,
            related_experiences=[e.id for e in contradictions],
            actionable_insights=["Update strategy based on new information"],
        )

    # Check if this is a first-time action
    action = partial_experience.get("action")
    exp_type = partial_experience.get("type")
    is_first_time = not any(e.action == action for e in recent_experiences)
    if is_first_time and exp_type == ExperienceType.SUCCESS:
        return ExperienceAnalysisResult(
            is_significant=True,
            learning=f"Successfully completed new action: {action}",
            confidence=0.7,
            actionable_insights=[f"{action} is now a known capability"],
        )

    # Check for failure patterns
    if exp_type == ExperienceType.FAILURE:
        failure_pattern = _detect_failure_pattern(partial_experience, recent_experiences)
        if failure_pattern:
            return ExperienceAnalysisResult(
                is_significant=True,
                learning=failure_pattern.learning,
                confidence=0.9,
                related_experiences=failure_pattern.related_ids,
                actionable_insights=failure_pattern.insights,
            )

    # Default: Record if confidence is high enough
    return ExperienceAnalysisResult(
        is_significant=(exp_type != ExperienceType.SUCCESS or random.random() > 0.7),
        confidence=0.5,
    )


def _find_similar_experiences(
    partial: dict[str, object],
    experiences: list[Experience],
) -> list[Experience]:
    return [
        e
        for e in experiences
        if (
            e.action == partial.get("action")
            and e.type == partial.get("type")
            and _similar_context(e.context, str(partial.get("context", "")))
        )
    ]


def _find_contradictions(
    partial: dict[str, object],
    experiences: list[Experience],
) -> list[Experience]:
    return [
        e
        for e in experiences
        if (
            e.action == partial.get("action")
            and e.context == partial.get("context")
            and e.type != partial.get("type")
        )
    ]


def _similar_context(context1: str, context2: str) -> bool:
    """Simple similarity check based on word overlap."""
    words1 = context1.lower().split()
    words2 = context2.lower().split()
    if not words1 or not words2:
        return False
    common_words = [w for w in words1 if w in words2]
    return len(common_words) / max(len(words1), len(words2)) > 0.5


def _detect_failure_pattern(
    partial: dict[str, object],
    experiences: list[Experience],
) -> FailurePattern | None:
    recent_failures = [e for e in experiences if e.type == ExperienceType.FAILURE][:10]
    action = partial.get("action")

    # Check for repeated failures
    same_action_failures = [e for e in recent_failures if e.action == action]
    if len(same_action_failures) >= 3:
        return FailurePattern(
            learning=(
                f"Action {action} has failed {len(same_action_failures)} times recently. "
                "Need alternative approach."
            ),
            related_ids=[e.id for e in same_action_failures],
            insights=[
                f"Avoid {action} until root cause is addressed",
                "Consider alternative actions to achieve the same goal",
            ],
        )

    # Check for cascading failures
    if len(recent_failures) >= 5:
        return FailurePattern(
            learning="Multiple consecutive failures detected. System may be in unstable state.",
            related_ids=[e.id for e in recent_failures[:5]],
            insights=[
                "Pause and reassess current approach",
                "Check system health and dependencies",
            ],
        )

    return None


async def detect_patterns(experiences: list[Experience]) -> list[DetectedPattern]:
    patterns: list[DetectedPattern] = []

    # Group experiences by action
    action_groups: dict[str, list[Experience]] = {}
    for exp in experiences:
        action_groups.setdefault(exp.action, []).append(exp)

    # Detect success/failure patterns
    for action, group in action_groups.items():
        positive_count = sum(1 for e in group if e.outcome == OutcomeType.POSITIVE)
        success_rate = positive_count / len(group) if group else 0

        if len(group) >= 5:
            if success_rate < 0.3:
                patterns.append(
                    DetectedPattern(
                        description=f"Action {action} has low success rate ({round(success_rate * 100)}%)",
                        frequency=len(group),
                        experiences=[e.id for e in group],
                        significance="high",
                    )
                )
            elif success_rate > 0.9:
                patterns.append(
                    DetectedPattern(
                        description=f"Action {action} is highly reliable ({round(success_rate * 100)}% success)",
                        frequency=len(group),
                        experiences=[e.id for e in group],
                        significance="medium",
                    )
                )

    # Detect time-based patterns
    hourly_groups = _group_by_hour(experiences)
    for hour, group in hourly_groups.items():
        if len(group) >= 10:
            negative_count = sum(1 for e in group if e.outcome == OutcomeType.NEGATIVE)
            failure_rate = negative_count / len(group)
            if failure_rate > 0.5:
                patterns.append(
                    DetectedPattern(
                        description=f"Higher failure rate during hour {hour} ({round(failure_rate * 100)}%)",
                        frequency=len(group),
                        experiences=[e.id for e in group[:5]],
                        significance="medium",
                    )
                )

    # Detect learning velocity
    learning_experiences = [
        e for e in experiences if e.type in (ExperienceType.DISCOVERY, ExperienceType.LEARNING)
    ]

    if len(learning_experiences) >= 3:
        recent_learning = learning_experiences[:10]
        time_diffs: list[int] = []
        for i in range(1, len(recent_learning)):
            prev = recent_learning[i - 1]
            curr = recent_learning[i]
            time_diffs.append(abs(prev.created_at - curr.created_at))

        if time_diffs:
            avg_time_between = sum(time_diffs) / len(time_diffs)
            patterns.append(
                DetectedPattern(
                    description=(
                        f"Learning new things every {round(avg_time_between / 60000)} "
                        "minutes on average"
                    ),
                    frequency=len(learning_experiences),
                    experiences=[e.id for e in recent_learning],
                    significance="medium",
                )
            )

    return patterns


def _group_by_hour(experiences: list[Experience]) -> dict[int, list[Experience]]:
    groups: dict[int, list[Experience]] = {}
    for exp in experiences:
        hour = datetime.fromtimestamp(exp.created_at / 1000, tz=UTC).hour
        groups.setdefault(hour, []).append(exp)
    return groups

File: packages/typescript/src/features/advanced-capabilities/experience/utils/experienceAnalyzer.ts
import type { Experience } from "../types";
import { ExperienceType, OutcomeType } from "../types";

export interface ExperienceAnalysis {
	isSignificant: boolean;
	learning?: string;
	confidence: number;
	relatedExperiences?: string[];
	actionableInsights?: string[];
}

export async function analyzeExperience(
	partialExperience: Partial<Experience>,
	recentExperiences: Experience[],
): Promise<ExperienceAnalysis> {
	// Check if this experience represents something new or significant
	const similarExperiences = findSimilarExperiences(
		partialExperience,
		recentExperiences,
	);

	// If we've seen this exact pattern many times, it's less significant
	if (similarExperiences.length > 5) {
		return {
			isSignificant: false,
			confidence: 0.3,
		};
	}

	// Check for contradictions with previous experiences
	const contradictions = findContradictions(
		partialExperience,
		recentExperiences,
	);
	if (contradictions.length > 0) {
		const firstContradiction = contradictions[0];
		if (firstContradiction) {
			const currentResult = partialExperience.result ?? "unknown result";
			const previousResult = firstContradiction.result ?? "unknown result";

			return {
				isSignificant: true,
				learning: `New outcome contradicts previous experience: ${currentResult} vs ${previousResult}`,
				confidence: 0.8,
				relatedExperiences: contradictions.map((e) => e.id),
				actionableInsights: ["Update strategy based on new information"],
			};
		}
	}

	// Check if this is a first-time action
	const isFirstTime = !recentExperiences.some(
		(e) => e.action === partialExperience.action,
	);
	if (isFirstTime && partialExperience.type === ExperienceType.SUCCESS) {
		return {
			isSignificant: true,
			learning: `Successfully completed new action: ${partialExperience.action}`,
			confidence: 0.7,
			actionableInsights: [
				`${partialExperience.action} is now a known capability`,
			],
		};
	}

	// Check for failure patterns
	if (partialExperience.type === ExperienceType.FAILURE) {
		const failurePattern = detectFailurePattern(
			partialExperience,
			recentExperiences,
		);
		if (failurePattern) {
			return {
				isSignificant: true,
				learning: failurePattern.learning,
				confidence: 0.9,
				relatedExperiences: failurePattern.relatedIds,
				actionableInsights: failurePattern.insights,
			};
		}
	}

	// Default: Record if confidence is high enough
	return {
		isSignificant:
			partialExperience.type !== ExperienceType.SUCCESS || Math.random() > 0.7,
		confidence: 0.5,
	};
}

function findSimilarExperiences(
	partial: Partial<Experience>,
	experiences: Experience[],
): Experience[] {
	return experiences.filter(
		(e) =>
			e.action === partial.action &&
			e.type === partial.type &&
			similarContext(e.context, partial.context || ""),
	);
}

function findContradictions(
	partial: Partial<Experience>,
	experiences: Experience[],
): Experience[] {
	return experiences.filter(
		(e) =>
			e.action === partial.action &&
			e.context === partial.context &&
			e.type !== partial.type,
	);
}

function similarContext(context1: string, context2: string): boolean {
	// Simple similarity check - could be enhanced with better NLP
	const words1 = context1.toLowerCase().split(/\s+/);
	const words2 = context2.toLowerCase().split(/\s+/);
	const commonWords = words1.filter((w) => words2.includes(w));
	return commonWords.length / Math.max(words1.length, words2.length) > 0.5;
}

interface FailurePattern {
	learning: string;
	relatedIds: string[];
	insights: string[];
}

function detectFailurePattern(
	partial: Partial<Experience>,
	experiences: Experience[],
): FailurePattern | null {
	const recentFailures = experiences
		.filter((e) => e.type === ExperienceType.FAILURE)
		.slice(0, 10);

	// Check for repeated failures
	const sameActionFailures = recentFailures.filter(
		(e) => e.action === partial.action,
	);
	if (sameActionFailures.length >= 3) {
		return {
			learning: `Action ${partial.action} has failed ${sameActionFailures.length} times recently. Need alternative approach.`,
			relatedIds: sameActionFailures.map((e) => e.id),
			insights: [
				`Avoid ${partial.action} until root cause is addressed`,
				"Consider alternative actions to achieve the same goal",
			],
		};
	}

	// Check for cascading failures
	if (recentFailures.length >= 5) {
		return {
			learning:
				"Multiple consecutive failures detected. System may be in unstable state.",
			relatedIds: recentFailures.slice(0, 5).map((e) => e.id),
			insights: [
				"Pause and reassess current approach",
				"Check system health and dependencies",
			],
		};
	}

	return null;
}

export async function detectPatterns(experiences: Experience[]): Promise<
	Array<{
		description: string;
		frequency: number;
		experiences: string[];
		significance: "low" | "medium" | "high";
	}>
> {
	const patterns: Array<{
		description: string;
		frequency: number;
		experiences: string[];
		significance: "low" | "medium" | "high";
	}> = [];

	// Group experiences by action
	const actionGroups = new Map<string, Experience[]>();
	experiences.forEach((exp) => {
		const group = actionGroups.get(exp.action) || [];
		group.push(exp);
		actionGroups.set(exp.action, group);
	});

	// Detect success/failure patterns
	actionGroups.forEach((group, action) => {
		const successRate =
			group.filter((e) => e.outcome === OutcomeType.POSITIVE).length /
			group.length;

		if (group.length >= 5) {
			if (successRate < 0.3) {
				patterns.push({
					description: `Action ${action} has low success rate (${Math.round(successRate * 100)}%)`,
					frequency: group.length,
					experiences: group.map((e) => e.id),
					significance: "high",
				});
			} else if (successRate > 0.9) {
				patterns.push({
					description: `Action ${action} is highly reliable (${Math.round(successRate * 100)}% success)`,
					frequency: group.length,
					experiences: group.map((e) => e.id),
					significance: "medium",
				});
			}
		}
	});

	// Detect time-based patterns
	const hourlyGroups = groupByHour(experiences);
	hourlyGroups.forEach((group, hour) => {
		if (group.length >= 10) {
			const failureRate =
				group.filter((e) => e.outcome === OutcomeType.NEGATIVE).length /
				group.length;
			if (failureRate > 0.5) {
				patterns.push({
					description: `Higher failure rate during hour ${hour} (${Math.round(failureRate * 100)}%)`,
					frequency: group.length,
					experiences: group.slice(0, 5).map((e) => e.id),
					significance: "medium",
				});
			}
		}
	});

	// Detect learning velocity
	const learningExperiences = experiences.filter(
		(e) =>
			e.type === ExperienceType.DISCOVERY || e.type === ExperienceType.LEARNING,
	);

	if (learningExperiences.length >= 3) {
		const recentLearning = learningExperiences.slice(0, 10);
		const timeDiffs: number[] = [];
		for (let i = 1; i < recentLearning.length; i++) {
			const previous = recentLearning[i - 1];
			const current = recentLearning[i];
			if (!previous || !current) {
				continue;
			}
			timeDiffs.push(previous.createdAt - current.createdAt);
		}
		const avgTimeBetweenLearning =
			timeDiffs.length > 0
				? timeDiffs.reduce((a, b) => a + b, 0) / timeDiffs.length
				: 0;

		patterns.push({
			description: `Learning new things every ${Math.round(avgTimeBetweenLearning / 60000)} minutes on average`,
			frequency: learningExperiences.length,
			experiences: recentLearning.map((e) => e.id),
			significance: "medium",
		});
	}

	return patterns;
}

function groupByHour(experiences: Experience[]): Map<number, Experience[]> {
	const groups = new Map<number, Experience[]>();

	experiences.forEach((exp) => {
		const hour = new Date(exp.createdAt).getHours();
		const group = groups.get(hour) || [];
		group.push(exp);
		groups.set(hour, group);
	});

	return groups;
}
